File: Algo/Leetcode/092ReverseLinkedListII.cpp
//1.Keep the original list. Keep the (m-1)-th Node place.
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* new_head = new ListNode(0);
        new_head->next = head;
        ListNode* res = new_head;
        int diff = right - left;
        while(left-->1){
            new_head = new_head->next;
            head = head->next;
        }

        while(diff-->0){
            ListNode* tmp = head->next;
            head->next = head->next->next;
            tmp->next = new_head->next;
            new_head->next = tmp;
        }
        return res->next;
    }
};

// find pre and succ, reverse l->r and reconnect
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if(left >= right) return head;
        ListNode* new_head = new ListNode(0);
        new_head->next = head;
        ListNode* l = new_head;
        ListNode* pre = NULL;
        while(left>0){
            pre = l;
            l = l->next;
            left--;
        }
        ListNode* r = new_head;
        ListNode* succ = NULL;
        while(right>=0){
            succ = r;
            r = r->next;
            right--;
        }
        swap(r, succ);

        r->next = NULL;
        pre->next = reverseList(l);
        while(l->next) l = l->next;
        l->next = succ;
        return new_head->next;
    }

    ListNode* reverseList(ListNode* cur) {
        ListNode* pre = NULL;
        ListNode* tmp;

        while(cur){
            tmp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = tmp;
        }
        return pre;
    }
};

File: Algo/Leetcode/092ReverseLinkedListII.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseBetween(self, head, left, right):
        """
        :type head: ListNode
        :type left: int
        :type right: int
        :rtype: ListNode
        """
        f = ListNode()
        f.next = head
        res = f
        t = ListNode()
        t.next = head
        # get the left end
        while left > 1:  # correct place
            f = f.next
            left -= 1
        # get the right end
        while right > 0:  # correct place
            t = t.next
            right -= 1
        # cut right
        _t = t.next
        t.next = None
        # direct connect
        if f.next is None:
            f.next = _t
            return head
        # reverse middle
        tmp = self.reverse(f.next)
        # get middle last
        last = tmp
        while last.next:
            last = last.next
        # connect middle last to right
        last.next = _t
        f.next = tmp

        return res.next

    def reverse(self, head):
        tail = head
        if head is None:
            return head

        while tail.next:
            tmp = tail.next
            tail.next = tail.next.next
            tmp.next = head
            head = tmp

        return head
